[SwiftUI] Find-in-Page integration (iOS)

rr-codes opened the pull request:
#### 31f53890413ab604b438db03c40f219b25f3ced1
<pre>
[SwiftUI] Find-in-Page integration (iOS)
<a href="https://bugs.webkit.org/show_bug.cgi?id=284715">https://bugs.webkit.org/show_bug.cgi?id=284715</a>
<a href="https://rdar.apple.com/141503364">rdar://141503364</a>

Reviewed by Wenson Hsieh.

Add support to enable &quot;Find in Page&quot; integration, on iOS only for now due to the many platform differences
between UIKit and AppKit.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
(EnvironmentValues.webViewFindContext):
(View.webViewFindNavigator(_:)):
(View.webViewFindDisabled(_:)):
(View.webViewReplaceDisabled(_:)):
(FindContext.isPresented):

Add several view modifiers which correspond to their native SwiftUI counterparts. The main modifier is
`webViewFindNavigator` which actually presents the find navigator when the binding value is true. The
`webViewFindDisabled` and `webViewReplaceDisabled` modifiers are used to control specific functionality
of the find navigator.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPageWebView.delegate):
(WebPageWebView.findInteraction(_:didBegin:)):
(WebPageWebView.findInteraction(_:didEnd:)):
(WebPageWebView.supportsTextReplacement):
(Delegate.findInteraction(_:didBegin:)):
(Delegate.findInteraction(_:didEnd:)):
(Delegate.supportsTextReplacement):

Add a delegate protocol to override and propogate some find interaction methods from WKWebView to WebView.

* Source/WebKit/UIProcess/API/Swift/WebView.swift:
(WebViewWrapper.findContext):
(WebViewWrapper.webView):
(WebViewWrapper.findInteraction(_:didBegin:)):
(WebViewWrapper.findInteraction(_:didEnd:)):
(WebViewWrapper.supportsTextReplacement):

Update the environment state&apos;s find context when the delegate methods are called.

(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeCoordinator):
(WebViewCoordinator.configuration):
(WebViewCoordinator.update(_:configuration:environment:)):
(WebViewCoordinator.updateFindInteraction(_:environment:)):

Present or dismiss the find navigator according to the find context state.

* Source/WebKit/UIProcess/Cocoa/Foundation+Extras.swift: Copied from Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift.
(onNextMainRunLoop(do:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Tools/SwiftBrowser/Source/ContentView.swift:
(ContentView.body):

Update SwiftBrowser to use this new functionality.

Canonical link: <a href="https://commits.webkit.org/288243@main">https://commits.webkit.org/288243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2cb05392ae1875e95cb69d6a8cabcdea2a1af8dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82224 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87356 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33285 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84329 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1959 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9768 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64081 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21825 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85293 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/1354 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74812 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44359 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1254 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28994 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32326 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29615 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88712 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9530 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6788 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72472 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9756 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70627 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71690 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15823 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12758 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9483 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9357 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12823 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11126 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
```diff
@@ -63,6 +65,27 @@ extension View {
+
+    @_spi(Private)
+    public func webViewReplaceDisabled(_ isDisabled: Bool = true) -> some View {
+        transformEnvironment(\.webViewFindContext) { $0.canReplace = !isDisabled }
```
I'd expect `canReplace` to key off of edibility, as well. Should that be reflected here?

rr-codes replied:
it's reflected here https://github.com/WebKit/WebKit/pull/37988/files#diff-c71e2a8d8a171cf1c377fcb17a2d78f09274e84b73e2453fda4f5ed689fd2decR49 ; if the page is not editable, replacement will never be available

axiixc commented on Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
```diff
@@ -63,6 +65,27 @@ extension View {
+
+    @_spi(Private)
+    public func webViewFindNavigator(isPresented: Binding<Bool>) -> some View {
+        environment(\.webViewFindContext, .init(isPresented: isPresented))
```
Does this introduce any sort of order dependency in the environment (from mixing `environment()` with `transformEnvironment()` like this)? If so, is that intentional? Ex, do these samples behave the same…
```swift
Content_A()
  .webViewFindNavigator(isPresented: $binding)
  .webViewFindDisabled(true)

Content_B()
  .webViewFindDisabled(true)
  .webViewFindNavigator(isPresented: $binding)
```

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -226,7 +264,7 @@ public class WebPage_v0 {
     }
 
-    private func createObservation<Value, BackingValue>(for keyPath: KeyPath<WebPage_v0, Value>, backedBy backingKeyPath: KeyPath<WKWebView, BackingValue>) -> NSKeyValueObservation {
+    private func createObservation<Value, BackingValue>(for keyPath: KeyPath<WebPage_v0, Value>, backedBy backingKeyPath: KeyPath<WebPageWebView, BackingValue>) -> NSKeyValueObservation {
```
You can probably get away not updating most of these. Unless you're planning to add bespoke properties then it's fine to just speak in terms of the super type.

rr-codes replied:
I initially tried so, but then it was complaining about converting mismatched KeyPath types

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -115,6 +140,57 @@ fileprivate struct WebViewRepresentable {
+
+        let isFindNavigatorVisible = findInteraction.isFindNavigatorVisible
+
+        // Showing or hiding the find navigator can change the first responder, which triggers a graph cycle if done synchronously.
```
Do we have a bug about this?

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -115,6 +140,57 @@ fileprivate struct WebViewRepresentable {
+        let isFindNavigatorVisible = findInteraction.isFindNavigatorVisible
+
+        // Showing or hiding the find navigator can change the first responder, which triggers a graph cycle if done synchronously.
+        if findContext.canFind, findContext.isPresented?.wrappedValue ?? false, !isFindNavigatorVisible {
```
IMO, don't use `,` in place of `&&`. Reserve it for variable bindings only.
```swift
if findContext.canFind && findContext.isPresented?.wrappedValue == true && !isFindNavigatorVisible {
```

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -115,6 +140,57 @@ fileprivate struct WebViewRepresentable {
+            onNextMainRunLoop {
+                findInteraction.presentFindNavigator(showingReplace: false)
+            }
+        } else if !(findContext.isPresented?.wrappedValue ?? true), isFindNavigatorVisible {
```
```swift
} else if findContext.isPresented?.wrappedValue == false && isFindNavigatorVisible {
```
I'd prefer direct comparisons to true and false, although the readability can be subjective.

axiixc commented on Source/WebKit/UIProcess/Cocoa/Foundation+Extras.swift:
```diff
@@ -0,0 +1,33 @@
+import Foundation
+
+@MainActor
+func onNextMainRunLoop(do body: @escaping @Sendable @MainActor () -> Void) {
```
I don't think you need `@Sendable` here since the closure is also `@MainActor`? Maybe you want `sending` instead, although all these keywords still confuse me too.

rr-codes committed:
[SwiftUI] Find-in-Page integration (iOS)
https://bugs.webkit.org/show_bug.cgi?id=284715
rdar://141503364

Reviewed by Wenson Hsieh.

Add support to enable "Find in Page" integration, on iOS only for now due to the many platform differences
between UIKit and AppKit.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
(EnvironmentValues.webViewFindContext):
(View.webViewFindNavigator(_:)):
(View.webViewFindDisabled(_:)):
(View.webViewReplaceDisabled(_:)):
(FindContext.isPresented):

Add several view modifiers which correspond to their native SwiftUI counterparts. The main modifier is
`webViewFindNavigator` which actually presents the find navigator when the binding value is true. The
`webViewFindDisabled` and `webViewReplaceDisabled` modifiers are used to control specific functionality
of the find navigator.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPageWebView.delegate):
(WebPageWebView.findInteraction(_:didBegin:)):
(WebPageWebView.findInteraction(_:didEnd:)):
(WebPageWebView.supportsTextReplacement):
(Delegate.findInteraction(_:didBegin:)):
(Delegate.findInteraction(_:didEnd:)):
(Delegate.supportsTextReplacement):

Add a delegate protocol to override and propogate some find interaction methods from WKWebView to WebView.

* Source/WebKit/UIProcess/API/Swift/WebView.swift:
(WebViewWrapper.findContext):
(WebViewWrapper.webView):
(WebViewWrapper.findInteraction(_:didBegin:)):
(WebViewWrapper.findInteraction(_:didEnd:)):
(WebViewWrapper.supportsTextReplacement):

Update the environment state's find context when the delegate methods are called.

(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeCoordinator):
(WebViewCoordinator.configuration):
(WebViewCoordinator.update(_:configuration:environment:)):
(WebViewCoordinator.updateFindInteraction(_:environment:)):

Present or dismiss the find navigator according to the find context state.

* Source/WebKit/UIProcess/Cocoa/Foundation+Extras.swift: Copied from Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift.
(onNextMainRunLoop(do:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Tools/SwiftBrowser/Source/ContentView.swift:
(ContentView.body):

Update SwiftBrowser to use this new functionality.

Canonical link: https://commits.webkit.org/288243@main
